src/common/FastRational.h

```diff
@@ -572,52 +599,75 @@ inline bool FastRational::isWellFormed() const
 {
     return (  wordPartValid() || mpqPartValid() )
            && ( !wordPartValid() || (den != 0 && gcd(absVal(num), den)==1) )
-           && ( !mpqPartValid()  || mpz_sgn(mpq_denref(mpq))!=0 );
+           && ( !mpqPartValid()  || mpz_sgn(mpq_denref(mpq))!=0 )
+           && ( !(wordPartValid() && mpqPartValid()) || wordAndMpqEqual());
+    // Check that if both wordPartValid() and mpqPartValid(), the are the same number
+}
+
+inline bool FastRational::wordAndMpqEqual() const {
+    assert(wordPartValid() && mpqPartValid());
+    uword int_den = mpz_get_ui(mpq_denref(mpq));
+    word int_num = mpz_get_si(mpq_numref(mpq));
+    return int_den == den and int_num == num;
 }
 
 inline FastRational::FastRational(word n, uword d) : state{State::WORD_VALID} {
     assert(d > 0);
     if (n == 0) {
         num = 0;
         den = 1;
-    } else if (n > 0) {
-        word common = gcd(uword(n), d);
-        num = n/common;
-        den = d/common;
     } else {
-        word common = gcd(uword(-n), d);
-        num = n/common;
-        den = d/common;
+        uword common = gcd<uword>(absVal(n), d);
+        if (common > 1) {
+            num = n / common;
+            den = d / common;
+        } else {
+            num = n;
+            den = d;
+        }
     }
 }
 
 inline void addition(FastRational& dst, const FastRational& a, const FastRational& b) {
     if (a.wordPartValid() && b.wordPartValid()) {
         if (b.num == 0) {
-            dst.num = a.num;
-            dst.den = a.den;
+            dst = a;
         } else if (a.num == 0) {
-            dst.num = b.num;
-            dst.den = b.den;
+            dst = b;
+        } else if (a == -b) {
+            dst.num = 0;
+            dst.den = 1;
```

Comment:
Shouldn't this behave as the other special cases? You get the same problem: what if `dst` has `mpq` part allocated?
Also I have to check, but I am suspicious about this check. What if `b.num` is WORD_MIN? This gets quite expensive. Maybe we could use `a.den == b.den and and b.num > WORD_MIN and a.num == -b.num`?
Now that I think about it, can we have a special case for `a.den == b.den`? The handling of that could be simpler, no?

Reply:
> Shouldn't this behave as the other special cases? You get the same problem: what if dst has mpq part allocated?

You're right.  I'll fix the others.

> Also I have to check, but I am suspicious about this check. What if b.num is WORD_MIN? This gets quite expensive. Maybe we could use a.den == b.den and and b.num > WORD_MIN and a.num == -b.num?

I micro benchmarked this (the benchmarking commit will be a separate PR as discussed), but it looks like this would be a good change.  The difference on my mac is (on small numbers):

 - Real(nom, den) + Real(-nom, den) 26.2 ns vs 16.5 ns
 - Real(nom, den) + Real(nom, den) 73.9 ns vs 64.6 ns

I get very similar results on mantella.  I'll do this for subtraction as well.

> Now that I think about it, can we have a special case for a.den == b.den? The handling of that could be simpler, no?

Could this have a snowball effect?  I suggest that we keep this in mind and get back to it on a future performance iteration.

Reply:
Agreed.